[question] ERROR: Package 'open62541/1.4.14' not resolved: Reference 'open62541/1.4.14#4f1ffaf9cf6e3a3d2366527bf4138a38%1764140285.608425' already exists.

### What is your question?

Hi. Can you help me figure this out?

I have a fork of conan-center-index and attached a CI-job to it to verify the packages that are relevant to us.
This job used to run fine, but it now fails with this error:

```
ERROR: Package 'open62541/1.4.14' not resolved: Reference 'open62541/1.4.14#4f1ffaf9cf6e3a3d2366527bf4138a38%1764140285.608425' already exists.
```

Last Friday I created a PR to sync our local fork with upstream. The run for the merge build succeeded, but the PR has not been completed (still pending review)
It started failing for runs after that.

This job runs on a on-prem computer, so there's caching between jobs. Could that be causing an issue when switching between CCI versions?
open62541/1.4.14 is a recipe with fixes I haven't pushed upstream yet.

> 
compiler.threads=mcf
compiler.version=14.2
os=Windows
[tool_requires]
open62541/*: cmake/[>=3.21 <4], ninja/[>=1.12.1 <2]
[conf]
tools.build:compiler_executables={'cpp': 'C:/msys64/mingw64/bin/g++.exe', 'c': 'C:/msys64/mingw64/bin/gcc.exe'}
tools.cmake.cmaketoolchain:generator=Ninja
tools.microsoft.bash:path=C:/msys64/usr/bin/bash.exe
tools.microsoft.bash:subsystem=msys2
[buildenv]
PATH=+(path)c:/msys64/usr/bin/


======== Computing dependency graph ========
xerces-c/3.3.0@_/_: Checking remote: cci-local
xerces-c/3.3.0@_/_: Checking remote: conancenter
Connecting to remote 'conancenter' anonymously
spdlog/1.15.3@_/_: Checking remote: cci-local
spdlog/1.15.3@_/_: Checking remote: conancenter
fmt/11.2.0: Checking remote: cci-local
fmt/11.2.0: Checking remote: conancenter
cmake/3.31.10: Checking remote: cci-local
cmake/3.31.10: Checking remote: conancenter
ninja/1.13.1: Checking remote: cci-local
ninja/1.13.1: Checking remote: conancenter
open62541pp/0.19.0: Checking remote: cci-local
open62541pp/0.19.0: Checking remote: conancenter
open62541/1.4.14: Checking remote: cci-local
open62541/1.4.14: Checking remote: conancenter
open62541/1.4.14: Retrieving from remote 'cci-local'...
Graph root
    conanfile.py: C:\agent\_work\3\s\conanfile.py
Requirements
    fmt/11.2.0#579bb2cdf4a7607621beea4eb4651e0f - Cache (Updated date) (cci-local)
    open62541pp/0.19.0#1960f53cac3131873651bb41363489b2 - Cache (Updated date) (cci-local)
    spdlog/1.15.3@_/_#3ca0e9e6b83af4d0151e26541d140c86 - Cache (conancenter)
    xerces-c/3.3.0@_/_#1460e1dcbd206b18a0ecf29f0b5ee361 - Cache (conancenter)
Build requirements
    cmake/3.31.10#313d16a1aa16bbdb2ca0792467214b76 - Cache (conancenter)
    ninja/1.13.1#294f8721dbcde145674f7ba44994700e - Cache (Updated date) (cci-local)
Resolved version ranges
    cmake/[>=3.21 <4]: cmake/3.31.10
    ninja/[>=1.12.1 <2]: ninja/1.13.1
Overrides
    open62541/1.4.6: ['open62541/1.4.14']
ERROR: Package 'open62541/1.4.14' not resolved: Reference 'open62541/1.4.14#4f1ffaf9cf6e3a3d2366527bf4138a38%1764140285.608425' already exists.
The system cannot find the path specified.
'cmake' is not recognized as an internal or external command,
operable program or batch file.
'cmake' is not recognized as an internal or external command,
operable program or batch file.
'cmake' is not recognized as an internal or external command,
operable program or batch file.
##[error]Cmd.exe exited with code '1'.
Finishing: Run PrintVersions to demonstrate working app with dependency

```
(venv) C:\agent\_work\3\s>conan list open62541/1.4.14:* -r cci-local
cci-local
  open62541
    open62541/1.4.14
      revisions
        4f1ffaf9cf6e3a3d2366527bf4138a38 (2025-11-26 07:02:42 UTC)
          packages

(venv) C:\agent\_work\3\s>conan --version
Conan version 2.23.0
```



### Have you read the CONTRIBUTING guide?

- [x] I've read the CONTRIBUTING guide

--- a/conan/internal/graph/proxy.py
+++ b/conan/internal/graph/proxy.py
@@ -3,7 +3,7 @@ from conan.internal.cache.conan_reference_layout import BasicLayout
 from conan.internal.graph.graph import (RECIPE_DOWNLOADED, RECIPE_INCACHE, RECIPE_NEWER,
                                         RECIPE_NOT_IN_REMOTE, RECIPE_UPDATED, RECIPE_EDITABLE,
                                         RECIPE_INCACHE_DATE_UPDATED, RECIPE_UPDATEABLE)
-from conan.internal.errors import NotFoundException
+from conan.internal.errors import NotFoundException, ConanReferenceAlreadyExistsInDB
 from conan.errors import ConanException
 
 
@@ -65,19 +65,22 @@ class ConanProxy:
             return recipe_layout, status, None
 
         # Something found in remotes, check if we already have the latest in local cache
-        # TODO: cache2.0 here if we already have a revision in the cache but we add the
-        #  --update argument and we find that same revision in server, we will not
-        #  download anything but we will UPDATE the date of that revision in the
-        #  local cache and WE ARE ALSO UPDATING THE REMOTE
-        #  Check if this is the flow we want to follow
         assert ref.timestamp
         cache_time = ref.timestamp
         if remote_ref.revision != ref.revision:
             if cache_time < remote_ref.timestamp:
                 # the remote one is newer
                 if should_update_reference(remote_ref, update):
-                    output.info("Retrieving from remote '%s'..." % remote.name)
-                    new_recipe_layout = self._download(remote_ref, remote)
+                    output.info(f"Updating to latest from remote '{remote.name}'...")
+                    try:
+                        new_recipe_layout = self._download(remote_ref, remote)
+                    except ConanReferenceAlreadyExistsInDB:
+                        # When updating to a newer revision in the server, but it already exists
+                        # in the cache with an older timestamp
+                        output.info(f"Latest from '{remote.name}' was found in "
+                                    "the cache, using it and updating its timestamp")
+                        new_recipe_layout = self._cache.recipe_layout(remote_ref)
+                        self._cache.update_recipe_timestamp(remote_ref)  # make it latest
                     status = RECIPE_UPDATED
                     return new_recipe_layout, status, remote
                 else:

--- a/test/integration/cache/cache2_update_test.py
+++ b/test/integration/cache/cache2_update_test.py
@@ -1,4 +1,6 @@
 import copy
+import json
+import textwrap
 from collections import OrderedDict
 
 import pytest
@@ -447,9 +449,9 @@ class TestUpdateFlows:
                                            ["libc", {"liba/1.0": "Cache",
                                                      "libb/1.0": "Cache"}],
                                            ["liba", {"liba/1.1": "Downloaded (default)",
-                                                       "libb/1.0": "Cache"}],
+                                                     "libb/1.0": "Cache"}],
                                            ["libb", {"liba/1.0": "Cache",
-                                                       "libb/1.1": "Downloaded (default)"}],
+                                                     "libb/1.1": "Downloaded (default)"}],
                                            ["", {"liba/1.0": "Cache",
                                                  "libb/1.0": "Cache"}],
                                            # Patterns not supported, only full name match
@@ -480,3 +482,49 @@ def test_muliref_update_pattern(update, result):
     tc.run(f'install --requires="liba/[>=1.0]" --requires="libb/[>=1.0]" -r default {update_flag}')
 
     tc.assert_listed_require(result)
+
+
+def test_update_remote_older_revision():
+    # https://github.com/conan-io/conan/issues/19313
+    c = TestClient(light=True, default_server_user=True)
+    zlib = textwrap.dedent("""
+       from conan import ConanFile
+       class Zlib(ConanFile):
+           name = "zlib"
+           version = "1.2.11"
+           exports_sources = "*"
+        """)
+    c.save({"conanfile.py": zlib,
+            "file.h": "//myheader"})
+    c.run("export")
+    c.run("upload * -r=default -c")
+
+    c2 = TestClient(light=True, servers=c.servers)
+    c2.run("graph info --requires=zlib/[*]")
+    rev1 = "2e87959c586811f8a4eaf12a327cc042"
+    c2.assert_listed_require({f"zlib/1.2.11#{rev1}": "Downloaded (default)"})
+
+    # Modify zlib code
+    c.save({"file.h": "//myheader 2"})
+    c.run("export")
+    c.run("upload * -r=default -c")
+
+    c2.run("graph info --requires=zlib/[*] --update")
+    rev2 = "934b3de03768d9030b61127d588d0a96"
+    c2.assert_listed_require({f"zlib/1.2.11#{rev2}": "Updated (default)"})
+
+    # revert this to old, make it latestlatest
+    c.save({"file.h": "//myheader"})
+    c.run("export")
+    c.run("remove * -r=default -c")
+    c.run("upload * -r=default -c")
+
+    c2.run("graph info --requires=zlib/[*] --update")
+    assert ("Latest from 'default' was found in "
+            "the cache, using it and updating its timestamp") in c2.out
+    c2.assert_listed_require({f"zlib/1.2.11#{rev1}": "Updated (default)"})
+
+    c2.run("list *#* --format=json")
+    revs = json.loads(c2.stdout)["Local Cache"]["zlib/1.2.11"]["revisions"]
+    # we check that the update from the server has made rev1 the latest in the cache again
+    assert revs[rev1]["timestamp"] > revs[rev2]["timestamp"]

--- a/test/integration/remote/test_local_recipes_index.py
+++ b/test/integration/remote/test_local_recipes_index.py
@@ -552,3 +552,48 @@ class TestResetRemote:
                 "requested, but it doesn't match the current available revision in source") in c.out
         assert ("ERROR: The 'zlib/1.2.11' package has 'exports_sources' but sources "
                 "not found in local cache") in c.out
+
+    def test_reverting_to_older_revision(self):
+        # https://github.com/conan-io/conan/issues/19313
+        folder = temp_folder()
+        recipes_folder = os.path.join(folder, "recipes")
+        zlib_config = textwrap.dedent("""
+           versions:
+             "1.2.11":
+               folder: all
+               """)
+        zlib = textwrap.dedent("""
+           from conan import ConanFile
+           class Zlib(ConanFile):
+               name = "zlib"
+               exports_sources = "*"
+               """)
+        save_files(recipes_folder,
+                   {"zlib/config.yml": zlib_config,
+                    "zlib/all/conanfile.py": zlib,
+                    "zlib/all/conandata.yml": "",
+                    "zlib/all/file.h": "//myheader"})
+
+        c = TestClient(light=True)
+        c.run(f"remote add local '{folder}'")
+        c.run("install --requires=zlib/[*] --build=missing")
+        rev1 = "169da4321a56b77e8538821613a81f1d"
+        c.assert_listed_require({f"zlib/1.2.11#{rev1}": "Downloaded (local)"})
+
+        # Modify zlib code
+        save_files(recipes_folder, {"zlib/all/file.h": "//myheader 222"})
+        c.run("install --requires=zlib/[*] --build=missing --update")
+        rev2 = "8e9fa314a3fd51ab4c930f7e4972f3e7"
+        c.assert_listed_require({f"zlib/1.2.11#{rev2}": "Updated (local)"})
+
+        # Modify zlib code again to a new one
+        save_files(recipes_folder, {"zlib/all/file.h": "//myheader 333"})
+        c.run("install --requires=zlib/[*] --update --build=missing")
+        rev3 = "dd54aebe11b96b3661d8360c68619a72"
+        c.assert_listed_require({f"zlib/1.2.11#{rev3}": "Updated (local)"})
+
+        # Revert to previous one
+        save_files(recipes_folder, {"zlib/all/file.h": "//myheader"})
+        c.run("install --requires=zlib/[*] --update --build=missing")
+        # This crashed https://github.com/conan-io/conan/issues/19313
+        c.assert_listed_require({f"zlib/1.2.11#{rev1}": "Updated (local)"})

